### PR TITLE
Fix table function clusterAllReplicas return wrong _shard_num

### LINKED ISSUE #21481: _shard_num and clusterAllReplicas()
**Bug description**
Unlike `cluster()`, `clusterAllReplicas()` does not implement `_shard_num`. 

I would expect
```
SELECT _shard_num
FROM clusterAllReplicas(prod, system.one)
ORDER BY _shard_num ASC
FORMAT PrettyCompactMonoBlock

┌─_shard_num─┐
│          0 │
│          0 │
│          0 │
│          0 │
│          0 │
│          0 │
│          0 │
│          0 │
│          0 │
│          0 │
│          0 │
│          0 │
│          0 │
│          0 │
│          0 │
│          0 │
│          0 │
│          0 │
│          0 │
│          0 │
│          0 │
│          0 │
│          0 │
│          0 │
│          0 │
│          0 │
│          0 │
│          0 │
│          0 │
│          0 │
└────────────┘
```
to have an output more useful like
```
SELECT _shard_num
FROM cluster(prod, system.one)
ORDER BY _shard_num ASC
FORMAT PrettyCompactMonoBlock

┌─_shard_num─┐
│          1 │
│          2 │
│          3 │
│          4 │
│          5 │
│          6 │
│          7 │
│          8 │
│          9 │
│         10 │
│         11 │
│         12 │
│         13 │
│         14 │
│         15 │
└────────────┘
```

**Releases where this has been tested**

* v20.3.12.112-stable

**References**

[cluster, clusterAllReplicas documentation](https://clickhouse.tech/docs/en/sql-reference/table-functions/cluster/)

[_shard_num documentation](https://clickhouse.tech/docs/en/engines/table-engines/special/distributed/#virtual-columns)

### PR DESCRIPTION
**Pull request by ucasfl**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):

- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Fix table function `clusterAllReplicas` returns wrong `_shard_num`. close #21481.

### CODE REVIEW

**Comment by CurtizJ** [comment]
Also, maybe it will be useful to have virtual column `_replica_num`.

File: src/Interpreters/Cluster.cpp [hunk 1]
```diff
@@ -573,6 +574,8 @@ Cluster::Cluster(Cluster::ReplicasAsShardsTag, const Cluster & from, const Setti
                 continue;   /// Duplicate host, skip.
 
             ShardInfo info;
+            info.shard_num = ++shard_num;
```
[the remark this answers is not in the document]
Comment (ucasFL) [reply src/Interpreters/Cluster.cpp:577]: I think it's correct. This function only used for `clusterAllReplicas` which regard all replicas as independent shards.

[the remark this answers is not in the document]
Comment (ucasFL) [reply src/Interpreters/Cluster.cpp:577]: It create a new `Cluster` contains all replicas as shard.

[the remark this answers is not in the document]
Comment (abyss7) [reply src/Interpreters/Cluster.cpp:577]: Let's just add functional test that shows that everything is correct for multi-replica shards.

[the remark this answers is not in the document]
Comment (ucasFL) [reply src/Interpreters/Cluster.cpp:577]: How can I add another cluster for test?

[the remark this answers is not in the document]
Comment (alexey-milovidov) [reply src/Interpreters/Cluster.cpp:577]: @ucasFL You can edit `tests/config.d/cluster.xml`

[the remark this answers is not in the document]
Comment (ucasFL) [reply src/Interpreters/Cluster.cpp:577]: Why I add cluster definition into server-test.xml or cluster.xml, the test still also show that cluster not found? @alexey-milovidov

[the remark this answers is not in the document]
Comment (azat) [reply src/Interpreters/Cluster.cpp:577]: >Why I add cluster definition into server-test.xml or cluster.xml, the test still also show that cluster not found?

AFAIK server-test.xml is not used anywhere (except for arcadia maybe), if you will add your cluster to `tests/config/config.d/clusters.xml` it should work.

FWIW fast-test (and other checks) got this configs via tests/config/install.sh, it already includes `clusters.xml`, here is a snippet from fast-test log:
```
+ ln -sf /ClickHouse/tests/config/config.d/clusters.xml /fasttest-workspace/db-fasttest/config.d/
```

Previous variant (https://github.com/ClickHouse/ClickHouse/commit/f6e83b579cc2b9db37b0a6315ee5633eaa6cf86a) - https://clickhouse-test-reports.s3.yandex.net/21498/f6e83b579cc2b9db37b0a6315ee5633eaa6cf86a/fast_test.html#fail1 was failed because you set port to 59000, and this port is used only for arcadia tests AFAIR.

So to summarize you should add everything like in f6e83b579cc2b9db37b0a6315ee5633eaa6cf86a, but with default port.

**Review by CurtizJ — APPROVED** [review]
Ok

**Comment by CurtizJ** [comment]
Commit 9f3ca62 is shown in cahnges of this PR, but it's already in master. I think it's because of several force-pushes.
